Enable tls related protocols for picasso on <=KitKat devices

evermind-zz opened the pull request:
<!-- Hey there. Thank you so much for improving NewPipe, and filling out the details. Having roughly the same layout helps everyone considerably :)-->

#### What is it?
- [ ] Bugfix (user facing)
- [ ] Feature (user facing)
- [x] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
<!-- While bullet points are the norm in this section, feel free to write free-form text instead of a list -->
 Could not load images with Picasso. This problem only surfaced as I maintain a fork of NewPipe that supports the video platform Rumble and the thumbnails stopped to load. The reason is that NewPipe moved from UniversalImageLoader to Picasso. This PR fixes that. It might fix other platforms on those devices too.
  - enable TLSv1.1/1.2 for OkHttpClient.Builder for Picasso on <=KitKat devices that support those protocols but have not been enabled.
#### APK testing 
<!-- Use a new, meaningfully named branch. The name is used as a suffix for the app ID to allow installing and testing multiple versions of NewPipe, e.g. "commentfix", if your PR implements a bugfix for comments. (No names like "patch-0" and "feature-1".)  -->
<!-- Remove the following line if you directly link the APK created by the CI pipeline. Directly linking is preferred if you need to let users test.-->
The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR.

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

litetex reviewed (commented):
I think this should be solved/done by #7175.

Btw: [TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA](https://ciphersuite.info/cs/TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA/) and [TLS_RSA_WITH_AES_256_CBC_SHA](https://ciphersuite.info/cs/TLS_RSA_WITH_AES_256_CBC_SHA/) are considered weak.

evermind-zz commented:
> I think this should be solved/done by #7175.
> 
> Btw: [TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA](https://ciphersuite.info/cs/TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA/) and [TLS_RSA_WITH_AES_256_CBC_SHA](https://ciphersuite.info/cs/TLS_RSA_WITH_AES_256_CBC_SHA/) are considered weak.

thanks for pointing that out. I will try that the NetCipher route #7175

evermind-zz commented:
I've checked NetCipher and get it up and running -- but they do NOT enable TLSv1.1 and TLSv1.2. So NetCipher is no solution for me as the server sp.rmbl.ws only supports TLSv1.2. Maybe I miss something to enable it there. Any hints appreciated. Thank you

litetex commented:
Okay, I reviewed the PR again and I don't understand the problems at all:
As stated in the javadoc:
https://www.javadoc.io/doc/com.squareup.okhttp3/okhttp/3.3.1/okhttp3/OkHttpClient.Builder.html#sslSocketFactory-javax.net.ssl.SSLSocketFactory-javax.net.ssl.X509TrustManager-

> Sets the socket factory and trust manager used to secure HTTPS connections. If unset, the system defaults will be used. 

We already overwrite the system default on devices that run Android 4.4 (API 19). 
So in theory there should be no problems as TLS 1.1. and 1.2 are already enabled for Picasso.

I think the problem here is that [PicassoHelper gets initialized](https://github.com/TeamNewPipe/NewPipe/blob/b5ad24eb475c1d0280fc8094438643acc77bbb74/app/src/main/java/org/schabi/newpipe/App.java#L107) before the [TLS is configured](https://github.com/TeamNewPipe/NewPipe/blob/b5ad24eb475c1d0280fc8094438643acc77bbb74/app/src/main/java/org/schabi/newpipe/MainActivity.java#L134-L137).
What happens if you move https://github.com/TeamNewPipe/NewPipe/blob/b5ad24eb475c1d0280fc8094438643acc77bbb74/app/src/main/java/org/schabi/newpipe/MainActivity.java#L134-L137 to https://github.com/TeamNewPipe/NewPipe/blob/b5ad24eb475c1d0280fc8094438643acc77bbb74/app/src/main/java/org/schabi/newpipe/App.java#L91?

Does this fix your problems?

evermind-zz commented:
This was my first try and it did not work. Does OkHttp3 build on top of [HttpsURLConnection](https://developer.android.com/reference/javax/net/ssl/HttpsURLConnection#setDefaultSSLSocketFactory(javax.net.ssl.SSLSocketFactory))? As the documentation states:

> Sets the default SSLSocketFactory inherited by new instances of this class.
>
> The socket factories are used when creating sockets for secure https URL connections.
>
I verified the OkHttpClient object is not using the TLSSocketFactoryCompat class in PicassoHelper but OpenSSLSocketFactoryImpl. Only the protocols ["TLSv1", "SSLv3"] are enabled

So I am not sure if this setting helps at all:
https://github.com/TeamNewPipe/NewPipe/blob/b5ad24eb475c1d0280fc8094438643acc77bbb74/app/src/main/java/org/schabi/newpipe/MainActivity.java#L134-L137

For me also media.ccc.de is not showing pictures on Android 4 emulator without this PR

In DownloaderImpl the TLSSocketFactoryCompat and additional Ciphers are explicitly set for its OkHttpClient.Builder().

I changed the PR now to not duplicate code and moved the DownloaderImpl enableModernTLS() method to a helper class and reuse that code to configure the PicassoHelper OkHttpClient.Builder() instance in the same way.

evermind-zz commented:
I created a new pull request using NetCipher which solves my problems
